### PR TITLE
[MINSTALL-177] Streamline the plugin

### PR DESCRIPTION
**Pull request by cstamas**
Original plugin made hoops and loops, instead to perform what it needed to perform. Partly to blame this was unfinished state of MAT API (it was able to install project only).

Installing project is needed in InstallMojo, but InstallFileMojo was forced to make hoops and loops due this, as it was passed one file (and maybe pomFile), and it was forced to create "fake" project, decorate and fake setup it with all whistle and bells, only to get it via MAT to resolver that would "decompose" it back into set of artifacts needing a deploy. So it went this file-artifact-project-artifact route, that made all the logic fragile and overly complicated.

This PR completely reworks m-install-p making it (almost trivially) simple: it does what it needs to do, without any fuss, and does it in streamlined way: InstallMojo will create a list of artifacts out of project and pass it to repository system for deploy, while InstallFileMojo literally prepares just a deployment request,  nothing more. No fuss, no magic, no fake project building etc.

Note: the code in mojos may or may not need to be reusable, but definitely smells like some "Maven API-ish thing". 

Problems: InstallFileMojo implicitly implemented ID validation (thru fake project building), and it revealed the problem that Maven ID (groupId, artifactId) and version validation is deeply buried into maven-model-builder and is NOT reusable at all, hence a light copy of logic (rules for ID allowed characters and version forbidden characters) are copied over here.

---

https://issues.apache.org/jira/browse/MINSTALL-177

### CODE REVIEW

**Comment by cstamas** [comment]
Ping? Anyone?

File: src/main/java/org/apache/maven/plugins/install/InstallMojo.java [hunk 1]
```diff
@@ -153,31 +163,90 @@ private boolean allProjectsMarked()
...
         try
         {
-            installer.install( session.getProjectBuildingRequest(), new ProjectInstallerRequest().setProject( pir ) );
+            repositorySystem.install( session.getRepositorySession(), processProject( project ) );
         }
-        catch ( IOException e )
+        catch ( IllegalArgumentException e )
         {
-            throw new MojoFailureException( "IOException", e );
+            throw new MojoFailureException( e.getMessage(), e );
```
Comment (slawekjaranowski) [inline src/main/java/org/apache/maven/plugins/install/InstallMojo.java:174]: why not `MojoExecutionException`

I must find a discussion, but as I remember `MojoFailureException` should not be used ....

File: src/main/java/org/apache/maven/plugins/install/InstallFileMojo.java [hunk 2]
```diff
@@ -158,48 +165,36 @@
...
+            DefaultRepositorySystemSession newSession = new DefaultRepositorySystemSession(
+                    session.getRepositorySession() );
+            // Clear cache, since we're using a new local repository
+            newSession.setCache( new DefaultRepositoryCache() );
+            // keep same repositoryType
+            String contentType = newSession.getLocalRepository().getContentType();
+            if ( "enhanced".equals( contentType ) )
+            {
+                contentType = "default";
+            }
```
Comment (slawekjaranowski) [inline src/main/java/org/apache/maven/plugins/install/InstallFileMojo.java:192]: in comments `keep same repositoryType` but change from `enhanced` to `default`

Comment (cstamas) [reply src/main/java/org/apache/maven/plugins/install/InstallFileMojo.java:192]: This is resolver (actually one of oldest Sonatype Aether) thing: the "enhanced" repository was introduced after the "simple" repository and was promoted as default (so priority-wise both act on "default" content type, but enhanced has higher priority), and enhanced repo reports content type enhanced. Is mess and inconsistent, but this is like it due legacy (in resolver).
Refs:
https://github.com/apache/maven-resolver/blob/master/maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/EnhancedLocalRepositoryManagerFactory.java#L98
https://github.com/apache/maven-resolver/blob/master/maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/EnhancedLocalRepositoryManager.java#L79

File: src/main/java/org/apache/maven/plugins/install/InstallMojo.java (hunk 1 above)
Comment (cstamas) [reply src/main/java/org/apache/maven/plugins/install/InstallMojo.java:174]: Hm, we need to clean up this I guess.... (as now I see that I used the two inconsistently).

File: src/main/java/org/apache/maven/plugins/install/InstallFileMojo.java (hunk 2 above)
Comment (cstamas) [reply src/main/java/org/apache/maven/plugins/install/InstallFileMojo.java:192]: https://issues.apache.org/jira/browse/MRESOLVER-267

File: src/main/java/org/apache/maven/plugins/install/InstallMojo.java (hunk 1 above)
Comment (slawekjaranowski) [reply src/main/java/org/apache/maven/plugins/install/InstallMojo.java:174]: I've started discussion on dev list about it 😄

**Review by gnodet — APPROVED** [review]

**Comment by jira-importer** [comment]
Resolve #172

**Comment by jira-importer** [comment]
Resolve #172

**Pull request merged**